pkg/executor/adapter_slow_log.go

```diff
@@ -0,0 +1,183 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"bytes"
+	"context"
+	"strings"
+	"time"
+
+	"github.com/pingcap/tidb/pkg/keyspace"
+	"github.com/pingcap/tidb/pkg/parser/ast"
+	"github.com/pingcap/tidb/pkg/sessionctx/variable"
+	"github.com/tikv/client-go/v2/util"
+)
+
+func init() {
+	variable.SlowLogRuleFieldAccessors[strings.ToLower(variable.SlowLogPlanDigest)] = variable.SlowLogFieldAccessor{
+		Parse: variable.ParseString,
+		Setter: func(_ context.Context, seVars *variable.SessionVars, items *variable.SlowQueryLogItems) {
+			_, planDigest := GetPlanDigest(seVars.StmtCtx)
+			items.PlanDigest = planDigest.String()
+		},
+		Match: func(_ *variable.SessionVars, items *variable.SlowQueryLogItems, threshold any) bool {
+			return variable.MatchEqual(threshold, strings.ToLower(items.PlanDigest))
+		},
+	}
+}
+
+// PrepareSlowLogItemsForRules builds a SlowQueryLogItems containing only the fields referenced by current session's SlowLogRules.
+// These pre-collected fields are later used for matching SQL execution details against the rules.
+func PrepareSlowLogItemsForRules(ctx context.Context, seVars *variable.SessionVars) *variable.SlowQueryLogItems {
+	items := &variable.SlowQueryLogItems{}
+	for field := range seVars.SlowLogRules.AllConditionFields {
+		gs := variable.SlowLogRuleFieldAccessors[field]
+		if gs.Setter != nil {
+			gs.Setter(ctx, seVars, items)
+		}
+	}
+
+	return items
+}
+
+// Match checks whether the given SlowQueryLogItems satisfies any of the
+// current session's slow log trigger rules.
+//
+// Matching is evaluated in two levels of logical relationship:
+//   - Rules are combined with OR: if any rule is satisfied, the function returns true.
+//   - Conditions inside a rule are combined with AND: all conditions must be met for that rule.
+//
+// Returns true if any rule matches, false otherwise.
+func Match(seVars *variable.SessionVars, items *variable.SlowQueryLogItems) bool {
+	rules := seVars.SlowLogRules
+	// Or logical relationship
+	for _, rule := range rules.Rules {
+		matched := true
+
+		// And logical relationship
+		for _, condition := range rule.Conditions {
+			accessor := variable.SlowLogRuleFieldAccessors[strings.ToLower(condition.Field)]
+			if ok := accessor.Match(seVars, items, condition.Threshold); !ok {
+				matched = false
+				break
+			}
+		}
+		if matched {
+			return matched
+		}
+	}
+
+	return false
+}
+
+// CompleteSlowLogItemsForRules fills in the remaining SlowQueryLogItems fields
+// that are relevant to triggering the current session's SlowLogRules.
+// It's exporting for testing.
+func CompleteSlowLogItemsForRules(ctx context.Context, seVars *variable.SessionVars, items *variable.SlowQueryLogItems) {
```

Comment:
What's the difference between CompleteSlowLogItemsForRules and PrepareSlowLogItemsForRules

Reply:
PrepareSlowLogItemsForRules: The obtained object only contains the contents of the fields present in the rule; fields that are not present in the rule will not be processed. 
CompleteSlowLogItemsForRules will also obtain the fields whose values were not retrieved in PrepareSlowLogItemsForRules.